### PR TITLE
Plfm 2117

### PR DESCRIPTION
**Pull request by brucehoff**

### CODE REVIEW

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClient.java [hunk 1]
```diff
@@ -745,30 +746,189 @@ UserEvaluationPermissions getUserEvaluationPermissions(String evalId)
...
 	void deleteMembershipRequest(String requestId) throws SynapseException;
 
 
-	/**
-	 * Get the List of ColumnModels for TableEntity given the TableEntity's ID.
+	/**
+	 * Refesh the prefix-cache for retrieving teams and team members
+	 * @throws SynapseException
+	 */
+	void updateTeamSearchCache() throws SynapseException;
```
Comment (jay-hodgson) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClient.java:928]: Trying to think of when one should call this method.  Is the team search cache automatically updated when a team is created?
A: This is for integration testing only.  The service is locked down so only a Synapse admin may call it.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java [hunk 2]
```diff
@@ -264,20 +303,130 @@ public TeamMemberPair mapRow(ResultSet rs, int rowNum) throws SQLException {
...
 	@Override
-	public Map<TeamHeader, List<UserGroupHeader>> getAllTeamsAndMembers() throws DatastoreException {
-		List<TeamMemberPair> results = simpleJdbcTemplate.query(SELECT_ALL_TEAMS_AND_MEMBERS, teamMemberPairRowMapper);
-		Map<TeamHeader, List<UserGroupHeader>> map = new HashMap<TeamHeader, List<UserGroupHeader>>();
-		for (TeamMemberPair tmp : results) {
-			List<UserGroupHeader> ughs = map.get(tmp.getTeamHeader());
-			if (ughs==null) {
-				ughs = new ArrayList<UserGroupHeader>();
-				map.put(tmp.getTeamHeader(), ughs);
+	public Map<Team, Collection<TeamMember>> getAllTeamsAndMembers() throws DatastoreException {
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java:344]: make transactional

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java [hunk 3]
```diff
@@ -264,20 +303,130 @@ public TeamMemberPair mapRow(ResultSet rs, int rowNum) throws SQLException {
...
+				fillUserGroupHeaderFromUserProfileBlob(upProperties, ugh);
+			} else {
+				ugh.setIsIndividual(false);
+			}
+			return tm;
+		}
+	};
+	
+	@Override
+	public List<TeamMember> getMembersInRange(String teamId, long limit, long offset)
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java:402]: make transactional

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java [hunk 4]
```diff
@@ -264,20 +303,130 @@ public TeamMemberPair mapRow(ResultSet rs, int rowNum) throws SQLException {
...
+			if (members==null) throw new IllegalStateException("No members found for team ID: "+tmi.getTeamId());
+			TeamMember tm = members.get(tmi.getMemberId());
+			if (tm==null) throw new IllegalStateException("No member found for team ID: "+tmi.getTeamId()+", member ID: "+tmi.getMemberId());
+			tm.setIsAdmin(true);
+		}
+		Map<Team, Collection<TeamMember>> results = new HashMap<Team, Collection<TeamMember>>();
+		// finally, create the results to return
+		for (Long teamId : teamMap.keySet()) {
+			Team team = teamMap.get(teamId);
+			if (team==null) throw new IllegalStateException("Missing TeamHeader for team ID: "+teamId);
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java:373]: "team" not "TeamHeader"

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java [hunk 5]
```diff
@@ -264,20 +303,130 @@ public TeamMemberPair mapRow(ResultSet rs, int rowNum) throws SQLException {
...
+			Collection<TeamMember> teamMembers = teamMemberMap.get(teamId).values();
+			if (teamMembers.isEmpty()) throw new IllegalStateException("Missing team members for team ID :"+teamId);
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/DBOTeamDAOImpl.java:375]: if (teamMembers==null || teamMembers.isEmpty())

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 6]
```diff
@@ -386,14 +402,49 @@ public void updateACL(UserInfo userInfo, AccessControlList acl)
...
 
 	@Override
-	public Map<TeamHeader, List<UserGroupHeader>> getAllTeamsAndMembers()
+	public Map<Team, Collection<TeamMember>> getAllTeamsAndMembers()
 			throws DatastoreException {
 		return teamDAO.getAllTeamsAndMembers();
 	}
 
+	@Override
+	public void setPermissions(UserInfo userInfo, String teamId,
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java:416]: should be transactional

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/TeamServiceImpl.java [hunk 7]
```diff
@@ -49,29 +86,162 @@ public Team create(String userId, Team team) throws UnauthorizedException,
...
 	public PaginatedResults<Team> get(String fragment, long limit, long offset)
-			throws DatastoreException {
-		if (fragment==null) {
-			return teamManager.get(limit, offset);
-		} else {
-			// TODO fragment search
+			throws DatastoreException, NotFoundException  {
+		if (limit<1) throw new IllegalArgumentException("'limit' must be at least 1");
+		if (offset<0) throw new IllegalArgumentException("'offset' may not be negative");
+		if (fragment==null) return teamManager.get(limit, offset);
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/TeamServiceImpl.java:104]: if (fragment==null || fragment.trim().length()==0)

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/TeamServiceImpl.java [hunk 8]
```diff
@@ -49,29 +86,162 @@ public Team create(String userId, Team team) throws UnauthorizedException,
...
 			throws DatastoreException, NotFoundException {
-		// TODO Auto-generated method stub
-		return null;
+		
+		if (limit<1) throw new IllegalArgumentException("'limit' must be at least 1");
+		if (offset<0) throw new IllegalArgumentException("'offset' may not be negative");
+
+		// if there is no prefix provided, we just to a regular paginated query
+		// against the database and return the result.  We also clear out the private fields.
+		if (fragment==null) {
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/TeamServiceImpl.java:222]: || fragment.trim().length()==0

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/TeamServiceImpl.java [hunk 9]
```diff
@@ -49,29 +86,162 @@ public Team create(String userId, Team team) throws UnauthorizedException,
...
+		Trie<String, Collection<TeamMember>> teamSpecificMemberPrefixCache = teamMemberPrefixCache.get(teamId);
+		if (teamSpecificMemberPrefixCache==null) throw new NotFoundException("Unrecognized teamId: "+teamId);
+		
+		// Get the results from the cache
+		SortedMap<String, Collection<TeamMember>> matched = teamSpecificMemberPrefixCache.prefixMap(fragment.toLowerCase());
+		List<TeamMember> fullList = PrefixCacheHelper.flatten(matched, teamMemberComparator);
+		PaginatedResults<TeamMember> results = new PaginatedResults<TeamMember>();
+		results.setTotalNumberOfResults(fullList.size());
+		List<TeamMember> page = new ArrayList<TeamMember>((int)limit);
+		for (int i=(int)offset; i<offset+limit && i<fullList.size(); i++) page.add(fullList.get(i));
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/TeamServiceImpl.java:242]: write extractPage(List list, int limit, int offset) utility to be called in both places

File: services/repository/src/test/java/org/sagebionetworks/repo/web/service/TeamServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		ugh.setDisplayName("John Smith");
+		member.setMember(ugh);
+
+		Map<Team, Collection<TeamMember>> universe = new HashMap<Team, Collection<TeamMember>>();
+		universe.put(team, Arrays.asList(new TeamMember[]{member}));
+		when(mockTeamManager.getAllTeamsAndMembers()).thenReturn(universe);
+		teamService.setTeamManager(mockTeamManager);
+	}
+	
+	@Test
```
Comment (brucehoff) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/service/TeamServiceTest.java:50]: add tests in which no fragment is provided and verify that manager is called

**Pull request merged**